Allow matching on external_identifier for source/target contacts on Activity import

mattwire opened the pull request:
Overview
----------------------------------------
Allow matching on external_identifier for source/target contacts on Activity import.

Before
----------------------------------------
You can only match on source/target contact ID.

After
----------------------------------------
You can match source/target contacts using external_identifier.

Technical Details
----------------------------------------
Adds handling for source/target via external_indentifier in the same way that support for contact matching via external_identifier is handled.

Comments
----------------------------------------

mattwire committed:
Allow activity import matching source/target contact on external ID

demeritcowboy commented:
This mostly works but I'm confused about some things:

* What's the difference between target external id and the existing external id field? They both seem to do the same thing. Was the new one intended to be assignee?
* Also if you want to match on source external id, it still requires you to pick something for target, but then if you pick target contact (regular id not external id) it says mismatch when it tries to actually import. That doesn't happen if you pick regular contact ids for both source and target.

demeritcowboy commented:
@mattwire What do you want to do with this one?

mattwire commented on CRM/Import/Forms.php:
```diff
@@ -744,7 +744,7 @@ protected function getAvailableFields(): array {
       $name = str_replace('.', '__', $name);
-      $return[$name] = $field['html']['label'] ?? $field['title'];
+      $return[$name] = $field['title'];
     }
```
We sometimes modify the title to make it clearer - so always display it.

mattwire commented:
@demeritcowboy You were right.. also seems like a recent core update broke this patch. I've reworked and fixed it now.

external_identifier is/was the target contact (I've updated the label to make that clear) and this adds source_contact_external_identifier so we can map the source contact by external identifier as well.

demeritcowboy commented:
This looks good. This was the testing I used, where "ok-err" means you get an error as expected.

  |   |   | source |   |  
-- | -- | -- | -- | -- | --
  |   | none | contact id | external id | email
  | none | ok - err | ok - err | ok - err | n/a
target | contact id | ok | ok | ok | n/a
  | external id | ok | ok | ok | n/a
  | email | ok | ok | ok | n/a

There's another set of matrices that look like variations of this, one where the field combo matches to different contacts and one where they are the same contact, but I only spot-checked, and I'm not sure it's really a big deal since you'd probably only make choices like this by mistake. There's slightly different results for whether it's target or source. For target, it gives an error when the combos don't match. For source, it seems to continue and prefer the external id over contact id instead of telling you there's a mismatch. (The errors here are presented after import, as opposed to above where it's during preview.) Ideally it would always warn when there's a mismatch, but meh.

  |   |   | source
-- | -- | -- | --
  |   | none | contact id + external id
target | contact id + external id |   |  
  | contact id + email |   |  
  | external id + email |   |  

mattwire commented:
Thanks @demeritcowboy
